File: ultrasonic/src/main/java/org/moire/ultrasonic/fragment/NowPlayingFragment.java
package org.moire.ultrasonic.fragment;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.navigation.Navigation;

import org.moire.ultrasonic.R;
import org.moire.ultrasonic.domain.MusicDirectory;
import org.moire.ultrasonic.domain.PlayerState;
import org.moire.ultrasonic.service.DownloadFile;
import org.moire.ultrasonic.service.MediaPlayerController;
import org.moire.ultrasonic.subsonic.ImageLoaderProvider;
import org.moire.ultrasonic.util.Constants;
import org.moire.ultrasonic.util.NowPlayingEventDistributor;
import org.moire.ultrasonic.util.NowPlayingEventListener;
import org.moire.ultrasonic.util.Util;

import kotlin.Lazy;
import timber.log.Timber;

import static org.koin.java.KoinJavaComponent.inject;


/**
 * Contains the mini-now playing information box displayed at the bottom of the screen
 */
public class NowPlayingFragment extends Fragment {

    private static final int MIN_DISTANCE = 30;
    private float downX;
    private float downY;
    ImageView playButton;
    ImageView nowPlayingAlbumArtImage;
    TextView nowPlayingTrack;
    TextView nowPlayingArtist;

    private final Lazy<MediaPlayerController> mediaPlayerControllerLazy = inject(MediaPlayerController.class);
    private final Lazy<ImageLoaderProvider> imageLoader = inject(ImageLoaderProvider.class);
    private final Lazy<NowPlayingEventDistributor> nowPlayingEventDistributor = inject(NowPlayingEventDistributor.class);
    private NowPlayingEventListener nowPlayingEventListener;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        Util.applyTheme(this.getContext());
        super.onCreate(savedInstanceState);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        return inflater.inflate(R.layout.now_playing, container, false);
    }

    @Override
    public void onViewCreated(@NonNull final View view, @Nullable Bundle savedInstanceState) {

        playButton = view.findViewById(R.id.now_playing_control_play);
        nowPlayingAlbumArtImage = view.findViewById(R.id.now_playing_image);
        nowPlayingTrack = view.findViewById(R.id.now_playing_trackname);
        nowPlayingArtist = view.findViewById(R.id.now_playing_artist);

        nowPlayingEventListener = new NowPlayingEventListener() {
            @Override
            public void onDismissNowPlaying() { }
            @Override
            public void onHideNowPlaying() { }
            @Override
            public void onShowNowPlaying() { update(); }
        };

        nowPlayingEventDistributor.getValue().subscribe(nowPlayingEventListener);
    }

    @Override
    public void onResume() {
        super.onResume();
        update();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        nowPlayingEventDistributor.getValue().unsubscribe(nowPlayingEventListener);
    }

    private void update() {
        try
        {
            PlayerState playerState = mediaPlayerControllerLazy.getValue().getPlayerState();
            if (playerState == PlayerState.PAUSED) {
                playButton.setImageDrawable(Util.getDrawableFromAttribute(getContext(), R.attr.media_play));
            } else if (playerState == PlayerState.STARTED) {
                playButton.setImageDrawable(Util.getDrawableFromAttribute(getContext(), R.attr.media_pause));
            }

            DownloadFile file = mediaPlayerControllerLazy.getValue().getCurrentPlaying();
            if (file != null) {
                final MusicDirectory.Entry song = file.getSong();
                String title = song.getTitle();
                String artist = song.getArtist();

                imageLoader.getValue().getImageLoader().loadImage(nowPlayingAlbumArtImage, song, false, Util.getNotificationImageSize(getContext()), false, true);
                nowPlayingTrack.setText(title);
                nowPlayingArtist.setText(artist);

                nowPlayingAlbumArtImage.setOnClickListener(v -> {
                    Bundle bundle = new Bundle();

                    if (Util.getShouldUseId3Tags()) {
                        bundle.putBoolean(Constants.INTENT_EXTRA_NAME_IS_ALBUM, true);
                        bundle.putString(Constants.INTENT_EXTRA_NAME_ID, song.getAlbumId());
                    } else {
                        bundle.putBoolean(Constants.INTENT_EXTRA_NAME_IS_ALBUM, false);
                        bundle.putString(Constants.INTENT_EXTRA_NAME_ID, song.getParent());
                    }

                    bundle.putString(Constants.INTENT_EXTRA_NAME_NAME, song.getAlbum());
                    bundle.putString(Constants.INTENT_EXTRA_NAME_NAME, song.getAlbum());
                    Navigation.findNavController(getActivity(), R.id.nav_host_fragment).navigate(R.id.trackCollectionFragment, bundle);
                });
            }

            getView().setOnTouchListener((v, event) -> handleOnTouch(event));

            // This empty onClickListener is necessary for the onTouchListener to work
            getView().setOnClickListener(v -> {});

            playButton.setOnClickListener(v -> mediaPlayerControllerLazy.getValue().togglePlayPause());
        }
        catch (Exception x) {
            Timber.w(x, "Failed to get notification cover art");
        }
    }

    private boolean handleOnTouch(MotionEvent event) {
        switch (event.getAction())
        {
            case MotionEvent.ACTION_DOWN:
            {
                downX = event.getX();
                downY = event.getY();
                return false;
            }
            case MotionEvent.ACTION_UP:
            {
                float upX = event.getX();
                float upY = event.getY();

                float deltaX = downX - upX;
                float deltaY = downY - upY;

                if (Math.abs(deltaX) > MIN_DISTANCE)
                {
                    // left or right
                    if (deltaX < 0)
                    {
                        mediaPlayerControllerLazy.getValue().previous();
                        return false;
                    }
                    if (deltaX > 0)
                    {
                        mediaPlayerControllerLazy.getValue().next();
                        return false;
                    }
                }
                else if (Math.abs(deltaY) > MIN_DISTANCE)
                {
                    if (deltaY < 0)
                    {
                        nowPlayingEventDistributor.getValue().raiseNowPlayingDismissedEvent();
                        return false;
                    }
                    if (deltaY > 0)
                    {
                        return false;
                    }
                }
                Navigation.findNavController(getActivity(), R.id.nav_host_fragment).navigate(R.id.playerFragment);
                return false;
            }
        }
        return false;
    }
}

File: ultrasonic/src/main/kotlin/org/moire/ultrasonic/fragment/NowPlayingFragment.kt
/*
 * NowPlayingFragment.kt
 * Copyright (C) 2009-2021 Ultrasonic developers
 *
 * Distributed under terms of the GNU GPLv3 license.
 */

package org.moire.ultrasonic.fragment

import android.annotation.SuppressLint
import android.os.Bundle
import android.view.LayoutInflater
import android.view.MotionEvent
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.fragment.app.Fragment
import androidx.navigation.Navigation
import io.reactivex.rxjava3.disposables.Disposable
import java.lang.Exception
import kotlin.math.abs
import org.koin.android.ext.android.inject
import org.moire.ultrasonic.R
import org.moire.ultrasonic.domain.PlayerState
import org.moire.ultrasonic.service.MediaPlayerController
import org.moire.ultrasonic.service.RxBus
import org.moire.ultrasonic.subsonic.ImageLoaderProvider
import org.moire.ultrasonic.util.Constants
import org.moire.ultrasonic.util.Settings
import org.moire.ultrasonic.util.Util.applyTheme
import org.moire.ultrasonic.util.Util.getDrawableFromAttribute
import org.moire.ultrasonic.util.Util.getNotificationImageSize
import timber.log.Timber

/**
 * Contains the mini-now playing information box displayed at the bottom of the screen
 */
@Suppress("unused")
class NowPlayingFragment : Fragment() {

    private var downX = 0f
    private var downY = 0f

    private var playButton: ImageView? = null
    private var nowPlayingAlbumArtImage: ImageView? = null
    private var nowPlayingTrack: TextView? = null
    private var nowPlayingArtist: TextView? = null

    private var playerStateSubscription: Disposable? = null
    private val mediaPlayerController: MediaPlayerController by inject()
    private val imageLoader: ImageLoaderProvider by inject()

    override fun onCreate(savedInstanceState: Bundle?) {
        applyTheme(this.context)
        super.onCreate(savedInstanceState)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        return inflater.inflate(R.layout.now_playing, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        playButton = view.findViewById(R.id.now_playing_control_play)
        nowPlayingAlbumArtImage = view.findViewById(R.id.now_playing_image)
        nowPlayingTrack = view.findViewById(R.id.now_playing_trackname)
        nowPlayingArtist = view.findViewById(R.id.now_playing_artist)
        playerStateSubscription =
            RxBus.playerStateObservable.subscribe { update() }
    }

    override fun onResume() {
        super.onResume()
        update()
    }

    override fun onDestroy() {
        super.onDestroy()
        playerStateSubscription!!.dispose()
    }

    @SuppressLint("ClickableViewAccessibility")
    private fun update() {
        try {
            val playerState = mediaPlayerController.playerState

            if (playerState === PlayerState.PAUSED) {
                playButton!!.setImageDrawable(
                    getDrawableFromAttribute(
                        requireContext(), R.attr.media_play
                    )
                )
            } else if (playerState === PlayerState.STARTED) {
                playButton!!.setImageDrawable(
                    getDrawableFromAttribute(
                        requireContext(), R.attr.media_pause
                    )
                )
            }

            val file = mediaPlayerController.currentPlaying

            if (file != null) {
                val song = file.song
                val title = song.title
                val artist = song.artist

                imageLoader.getImageLoader().loadImage(
                    nowPlayingAlbumArtImage,
                    song,
                    false,
                    getNotificationImageSize(requireContext())
                )

                nowPlayingTrack!!.text = title
                nowPlayingArtist!!.text = artist

                nowPlayingAlbumArtImage!!.setOnClickListener {
                    val bundle = Bundle()

                    if (Settings.shouldUseId3Tags) {
                        bundle.putBoolean(Constants.INTENT_EXTRA_NAME_IS_ALBUM, true)
                        bundle.putString(Constants.INTENT_EXTRA_NAME_ID, song.albumId)
                    } else {
                        bundle.putBoolean(Constants.INTENT_EXTRA_NAME_IS_ALBUM, false)
                        bundle.putString(Constants.INTENT_EXTRA_NAME_ID, song.parent)
                    }

                    bundle.putString(Constants.INTENT_EXTRA_NAME_NAME, song.album)
                    bundle.putString(Constants.INTENT_EXTRA_NAME_NAME, song.album)

                    Navigation.findNavController(requireActivity(), R.id.nav_host_fragment)
                        .navigate(R.id.trackCollectionFragment, bundle)
                }
            }
            requireView().setOnTouchListener { _: View?, event: MotionEvent ->
                handleOnTouch(event)
            }

            // This empty onClickListener is necessary for the onTouchListener to work
            requireView().setOnClickListener { }
            playButton!!.setOnClickListener { mediaPlayerController.togglePlayPause() }
        } catch (all: Exception) {
            Timber.w(all, "Failed to get notification cover art")
        }
    }

    private fun handleOnTouch(event: MotionEvent): Boolean {
        when (event.action) {
            MotionEvent.ACTION_DOWN -> {
                downX = event.x
                downY = event.y
            }

            MotionEvent.ACTION_UP -> {
                val upX = event.x
                val upY = event.y
                val deltaX = downX - upX
                val deltaY = downY - upY

                if (abs(deltaX) > MIN_DISTANCE) {
                    // left or right
                    if (deltaX < 0) {
                        mediaPlayerController.previous()
                    }
                    if (deltaX > 0) {
                        mediaPlayerController.next()
                    }
                } else if (abs(deltaY) > MIN_DISTANCE) {
                    if (deltaY < 0) {
                        RxBus.dismissNowPlayingCommandPublisher.onNext(Unit)
                    }
                } else {
                    Navigation.findNavController(requireActivity(), R.id.nav_host_fragment)
                        .navigate(R.id.playerFragment)
                }
            }
        }
        return false
    }

    companion object {
        private const val MIN_DISTANCE = 30
    }
}
